### PR TITLE
add computedPropertyDefinitions

### LINKED ISSUE #39: Enable mixing in "observable"-based decorators
I want to make it possible to easily mixin behaviors like async computes and the new resolver observables:

```js
ViewModel extends observe.Object {
  get todosPromise(){
    return Todo.getList({filter: this.filter})
  },
  @async
  todos(resolve){
    this.todosPromise.then(resolve)
  },
  @resolver
  requestCounter(prop){
    var count = 0;
    prop.resolve(count);
    prop.listenTo("todosPromise", function(){
      prop.resolve( ++count );
    })
  }
}
```

We need to be able to make it easy to mixin these observables in a performant way similar to `can-define`.  `can-define` is able to lazily construct observables for instances only when the instance property is actually bound, read or set.

As this is beyond what Object.defineProperty enables (only get and set, no binding lifecycle hooks), we need to make something ourselves.

Here's some background on how `can-define` works.  `can-define` keeps a `definitions` object on the prototype with a `computedInitializers` object:

```js
Person.prototype.definitions.computedInitializers = { ... }
```

A `computedInitializer` object looks like a object of property names to functions that return a `computeObj`: 

```js
{
  fullName() {
    return {
      // NOTICE how GETTER needs to be available in the scope
      compute: new Observation( GETTER, this),
      oldValue: undefined,  // cache the old value for performance reasons
      count: 0, // number of bindings
      handler: (newVal) => { // forward events to map
        var oldValue = computeObj.oldValue;
        computeObj.oldValue = newVal;
        this.dispatch({ type: prop, target: this }, [newVal, oldValue]); 
      }
    }
  }
}
```

`can-define` also creates a lazily-defined `_computed` property on the prototype (`~` denotes lazy).  When _computed is read for the first time, it will create create an object of lazily defined getters that return the  `computeObj` from above:

```
Person.prototype~_computed = function(){
  var instanceComputed = {};
  for(initializerFn of computedInitializers) {
    instanceComputed~[prop] = initializerFn.bind(this)
  }
  return instanceComputed;
}
```

> NOTE: the lazily defined computed property is always read during initialization of the define map instance.  I think this is to work better with sealed instances, but I'm not 100% sure.  This means that it really doesn't need to written as lazy.  We could have had a `this._computed = this.initializeComputed()` sort of thing in the constructor.    

Putting this another way, when an instance is created, it does some partial work to make a `_computed` object on the instance that lazily creates the `computedObj`. The partial work is really just making sure `fullName` above will be called with the right `this`.

Finally, `can-define` just looks in `_computed` when binding and unbinding.

## Making this useful

We want nice APIs that people can hook into.  I'll start with rough proposals that work backwards in terms of performance:

### installing an already instantiated observable value as a property on an instance

```js
var obs = observe({});
obs[can.computed].fullName = {count, handler, compute, oldValue}
```

> NOTES:
> - compute should probably be called `observable`
> - ideally we could just set the observable and everything else would happen auto-magically. Perhaps the `computed` object should be a proxy itself?  Scary.

### lazily installing observable property behavior on an instance 

```js
var obs = observe({});
defineLazyValue(obs[can.computed],"fullName", function(){
  return {count, handler, compute: new Observation( ..., this ), oldValue}
})
```

### lazily installing observable property behavior on all instances via the prototype

```js
Person.prototype[can.computedDefinitions].fullName = function(){
  return {count, handler, compute: new Observation( ..., this ), oldValue}
}
```

### PR DESCRIPTION
**Pull request by christopherjbaker**
fixes #39

### CODE REVIEW

File: doc/can-observe.md [hunk 1]
```diff
@@ -309,6 +309,54 @@ class WidgetViewModel {
 }
 ```
 
+## Extending can-observe with Observables
```
Comment (justinbmeyer) [inline doc/can-observe.md]: We need a better term than `Observables` ... something like `with rich property behavior`.

Comment (justinbmeyer) [reply doc/can-observe.md]: also it couldn't hurt to explain what the goal is here before jumping into technical details.

File: doc/can-observe.md [hunk 2]
```diff
@@ -309,6 +309,54 @@ class WidgetViewModel {
...
+
+can-observe recognizes a `can.computedPropertyDefinitions` property; getting or setting a key on the object that matches a key in the `can.computedPropertyDefinitions` object will use this functionality. Specifically, the values in this object are functions which return an observation (typically with [can-observation]). The first time one of these properties is accessed, the function is run, and the observation is cached, to be used for all future use on _that instance_.
```
Comment (justinbmeyer) [inline doc/can-observe.md]: >  getting or setting a key on the object

Which object?  The `proxy-wrapped` object.  I think I might call them proxy-wrapped elsewhere, if not, you might want to establish this term.

File: doc/can-observe.md [hunk 3]
```diff
@@ -309,6 +309,54 @@ class WidgetViewModel {
...
+
+### Defining Your Own
+
+We have tried to make it as easy as possible to define your own extensions. To that end, there is a `addComputedPropertyDefinition` helper; it accepts the target object, the property name, and the observation generator function.
+
+> _Note_: this `target` can be an instance (such as the output of `observe({})`) *or* a class (such as `class extends ObserveObject`). If target is a class, it will put this method on the prototype, so all instances get this new behavior.
+
+```javascript
+var helpers = require("can-observe/object/helpers");
```
Comment (justinbmeyer) [inline doc/can-observe.md]: Why would this be `object/helpers`?  I suppose it's for `objects`, but it's not really for `ObserveObject` only (is it?).  Might need to think about where this helper belongs.

File: doc/can-observe.md (hunk 2 above)
Comment (justinbmeyer) [reply doc/can-observe.md]: is `can.computedPropertyDefinitions` formally described somewhere?  We might want to do that.

Comment (christopherjbaker) [reply doc/can-observe.md]: Where would this formal definition go? in can-observe? in can-symbols?

File: doc/can-observe.md (hunk 3 above)
Comment (christopherjbaker) [reply doc/can-observe.md]: Like the decorators, the helpers will (i think) primarily be used with Object, though they can also be used with Array, and even with the return value of `observe(...)` regardless of its input type. The helpers could certainly be moved up a level, at `can-observe/helpers`, so they aren't suggested to be connected to Object.

I'm a little torn on location for the decorators. `can-define` (and some other packages) have a `can-define-X` pattern (`can-define-stream`, `can-define-validators`, etc). I imagine can-observe would follow the same structure, though I'm not sure where the 2 current decorators would live in that case. They could stay where they are (moved up with helpers, presumably), but then how do we decide when a new decorator goes here or in a separate package?

File: doc/can-observe.md (hunk 2 above)
Comment (justinbmeyer) [reply doc/can-observe.md]: in can-observe

File: decorators-test.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	person.value = "changed";
+	count++;
+	person.value = "again";
+
+	QUnit.equal(didRun, true, "on(count) was run");
+});
+
+
+
+function testDecoratorGetter(decoratorName, decorator, propName, getter, tester) {
```
Comment (justinbmeyer) [inline decorators-test.js]: move these to the top or another file

File: decorators-test.js [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	});
+}
+
+function testDecoratorMethod(decoratorName, decorator, propName, method, tester) {
+	if(classSupport) {
+		QUnit.test(decoratorName + " decorator with class Object prototype", function() {
+			var ran = false;
+
+			class Type extends ObserveObject {
+				[propName](resolve) {
```
Comment (justinbmeyer) [inline decorators-test.js]: is this supported everywhere class is supported?

Comment (christopherjbaker) [reply decorators-test.js]: yes, according to https://kangax.github.io/compat-table/es6/

**Pull request merged**